src/otx/core/model/module/base.py

```diff
@@ -0,0 +1,108 @@
+
+        # this line allows to access init params with 'self.hparams' attribute
+        # also ensures init params will be stored in ckpt
+        self.save_hyperparameters(logger=False, ignore=["otx_model"])
+
+    def training_step(self, inputs: OTXBatchDataEntity, batch_idx: int) -> Tensor:
+        """Step for model training."""
+        train_loss = self.model(inputs)
+
+        if isinstance(train_loss, Tensor):
+            self.log(
+                "train/loss",
+                train_loss,
+                on_step=True,
+                on_epoch=False,
+                prog_bar=True,
+            )
+            return train_loss
+        if isinstance(train_loss, dict):
+            for k, v in train_loss.items():
+                self.log(
+                    f"train/{k}",
+                    v,
+                    on_step=True,
+                    on_epoch=False,
+                    prog_bar=True,
+                )
+
+            train_loss = sum(train_loss.values())
+            self.log(
+                "train/loss",
+                train_loss,
+                on_step=True,
+                on_epoch=False,
+                prog_bar=True,
+            )
+            return train_loss
+
+        raise TypeError(train_loss)
+
+    def setup(self, stage: str) -> None:
+        """Lightning hook that is called at the beginning of fit (train + validate), validate, test, or predict.
+
+        This is a good hook when you need to build models dynamically or adjust something about
+        them. This hook is called on every process when using DDP.
+
+        :param stage: Either `"fit"`, `"validate"`, `"test"`, or `"predict"`.
+        """
+        if self.torch_compile and stage == "fit":
+            self.model = torch.compile(self.model)
+
+    def configure_optimizers(self) -> dict[str, Any]:
+        """Choose what optimizers and learning-rate schedulers to use in your optimization.
+
+        Normally you'd need one. But in the case of GANs or similar you might have multiple.
+
+        Examples:
+            https://lightning.ai/docs/pytorch/latest/common/lightning_module.html#configure-optimizers
+
+        :return: A dict containing the configured optimizers and learning-rate schedulers to be used for training.
+        """
+        optimizer = self.hparams.optimizer(params=self.parameters())
+        if self.hparams.scheduler is not None:
+            scheduler = self.hparams.scheduler(optimizer=optimizer)
+            return {
+                "optimizer": optimizer,
+                "lr_scheduler": {
+                    "scheduler": scheduler,
+                    "monitor": self.lr_scheduler_monitor_key,
+                    "interval": "epoch",
+                    "frequency": 1,
+                },
+            }
+        return {"optimizer": optimizer}
+
+    @property
+    def lr_scheduler_monitor_key(self) -> str:
+        """Metric name that the learning rate scheduler monitor."""
+        return "val/loss"
```

Comment:
For future consideration, it might be beneficial to develop additional interfaces for tasks such as deployment, optimization, and export.